### PR TITLE
Support array datatype

### PR DESCRIPTION
**Pull request by xiaocai2333**
issue: #26108 

1. Support a new data type, and element_type must be specified when constructing an Array field. 
  - Currently supported element_types include: `DataType.Boo`l, `DataType.Int8`, `DataType.Int16`, `DataType.Int32`, `DataType.Int64`, `DataType.Float`, `DataType.Double`, `DataType.Varchar`.
2. support new expression function: `array_lnegth(array)`
  - parameter `array` can be either `DataType.Array`, or the key of `DataType.Json`. If the value corresponding to the key is not an array, the length is considered to be 0.
3. support `array_contains`, `array_contains_any`, `array_contains_all`, semantics correspond to `json_contains`, `json_contains_any`, `json_contains_all`.

### CODE REVIEW

File: internal/core/src/common/Array.h [hunk 1]
```diff
@@ -0,0 +1,366 @@
...
+#include "fmt/core.h"
+#include "exceptions/EasyAssert.h"
+#include "Types.h"
+#include "FieldMeta.h"
+
+namespace milvus {
+
+class Array {
+ public:
+    Array() = default;
```
Comment (longjiquan) [inline internal/core/src/common/Array.h:32]: Remove this unless you need a default constructor.

File: internal/core/src/common/Array.h [hunk 2]
```diff
@@ -0,0 +1,366 @@
...
+            default: {
+                //                PanicInfo("unsupported element type for array");
+            }
+        }
+    }
+
+    Array(char* data,
+          size_t size,
+          DataType element_type,
+          std::vector<uint64_t> element_offsets)
```
Comment (longjiquan) [inline internal/core/src/common/Array.h:122]: Use && reference to avoid copy.

File: internal/core/src/common/Array.h [hunk 3]
```diff
@@ -0,0 +1,366 @@
...
+#include <arrow/array.h>
+#include <arrow/array/builder_primitive.h>
+#include "fmt/core.h"
+#include "exceptions/EasyAssert.h"
+#include "Types.h"
+#include "FieldMeta.h"
+
+namespace milvus {
+
+class Array {
```
Comment (longjiquan) [inline internal/core/src/common/Array.h:30]: Please implement the deconstructor and support the move semantics.

File: internal/core/src/common/Array.h [hunk 4]
```diff
@@ -0,0 +1,366 @@
...
+                                    ? size_ - offsets_.back()
+                                    : offsets_[index + 1] - offsets_[index];
+        if constexpr (std::is_same_v<T, std::string> ||
+                      std::is_same_v<T, std::string_view>) {
+            return T(data_ + offsets_[index], element_length);
+        }
+        return reinterpret_cast<T*>(data_)[index];
+    }
+
+    std::vector<uint64_t>
```
Comment (longjiquan) [inline internal/core/src/common/Array.h:147]: Please also provide an API like below:
```c++
const std::vector<uint64_t>&
get_immutable_offsets() const() {
    return offsets_;
}
```

**Review by longjiquan — CHANGES_REQUESTED** [review]

File: internal/core/src/mmap/Column.h [hunk 5]
```diff
@@ -286,10 +287,108 @@ class VariableColumn : public ColumnBase {
         views_.emplace_back(data_ + indices_.back(), size_ - indices_.back());
     }
 
- private:
```
Comment (longjiquan) [inline internal/core/src/mmap/Column.h:289]: Why we remove the `private`?

File: internal/core/src/mmap/Column.h [hunk 6]
```diff
@@ -286,10 +287,108 @@ class VariableColumn : public ColumnBase {
...
+           DataType element_type,
+           const std::vector<uint64_t>& element_indices) {
+        indices_.emplace_back(size_);
+        element_type_ = element_type;
+        element_indices_.emplace_back(element_indices);
+        ColumnBase::Append(data, size);
+    }
+
+    void
+    Seal(std::vector<uint64_t> indices = {},
```
Comment (longjiquan) [inline internal/core/src/mmap/Column.h:361]: Use && reference to avoid copy.

File: internal/core/src/mmap/Column.h [hunk 7]
```diff
@@ -286,10 +287,108 @@ class VariableColumn : public ColumnBase {
...
+    }
+
+    void
+    Append(const char* data, size_t size) {
+        indices_.emplace_back(size_);
+        ColumnBase::Append(data, size);
+    }
+
+    void
+    Append(const char* data,
```
Comment (yah01) [inline internal/core/src/mmap/Column.h:350]: As the array is a large type (e.g. Array of strings), providing a `Append` method that takes a `Array` would help if we move the array into the column

**Review by longjiquan — CHANGES_REQUESTED** [review]

File: internal/core/src/query/visitors/ExecExprVisitor.cpp [hunk 8]
```diff
@@ -501,6 +501,63 @@ ExecExprVisitor::ExecUnaryRangeVisitorDispatcher(UnaryRangeExpr& expr_raw)
     return ExecUnaryRangeVisitorDispatcherImpl<T>(expr_raw);
 }
 
+template <typename T>
+bool
+compareTwoJsonArray(T arr1, const proto::plan::Array& arr2) {
```
Comment (longjiquan) [inline internal/core/src/query/visitors/ExecExprVisitor.cpp:506]: Please add static assertion for T, so we can remember to add logic when new type is supported.

File: internal/core/src/common/Array.h [hunk 9]
```diff
@@ -0,0 +1,366 @@
...
+            }
+            case ScalarArray::kIntData: {
+                element_type_ = DataType::INT32;
+                length_ = field_data.int_data().data().size();
+                data_ = new char[length_ * sizeof(int)];
+                for (int i = 0; i < length_; ++i) {
+                    reinterpret_cast<int*>(data_)[i] =
+                        field_data.int_data().data(i);
+                    offsets_.emplace_back(size_);
+                    size_ += sizeof(int);
```
Comment (zhagnlu) [inline internal/core/src/common/Array.h:56]: this two line can simplify: offsets.emplace_back(i* sizeof(int));

File: internal/parser/planparserv2/parser_visitor.go [hunk 10]
```diff
@@ -992,25 +982,33 @@ func (v *ParserVisitor) getColumnInfoFromJSONIdentifier(identifier string) (*pla
...
+			}
 		} else if _, err := strconv.ParseInt(path, 10, 64); err != nil {
 			return nil, fmt.Errorf("json key must be enclosed in double quotes or single quotes: \"%s\"", path)
 		}
 		nestedPath = append(nestedPath, path)
 	}
 
 	if typeutil.IsJSONType(field.DataType) && len(nestedPath) == 0 {
-		return nil, fmt.Errorf("can not comparisons jsonField directly")
+		return nil, fmt.Errorf("can not comparisons json fields directly")
```
Comment (longjiquan) [inline internal/parser/planparserv2/parser_visitor.go:995]: Typo here. I guess `compare` should be used.

File: internal/parser/planparserv2/parser_visitor.go [hunk 11]
```diff
@@ -992,25 +982,33 @@ func (v *ParserVisitor) getColumnInfoFromJSONIdentifier(identifier string) (*pla
...
+	}
+
+	if typeutil.IsArrayType(field.DataType) && len(nestedPath) == 0 {
+		return nil, fmt.Errorf("can not comparisons array fields directly")
```
Comment (longjiquan) [inline internal/parser/planparserv2/parser_visitor.go:999]: Ditto.

File: internal/parser/planparserv2/utils.go [hunk 12]
```diff
@@ -331,9 +337,49 @@ func relationalCompatible(t1, t2 schemapb.DataType) bool {
...
+	if columnInfo := toColumnInfo(expr); columnInfo != nil {
+		return columnInfo.GetElementType()
+	}
+	if valueExpr := expr.expr.GetValueExpr(); valueExpr != nil {
+		return valueExpr.GetValue().GetArrayVal().GetElementType()
+	}
+	return schemapb.DataType_None
+}
+
+func canCompared(left, right *ExprWithType) bool {
```
Comment (longjiquan) [inline internal/parser/planparserv2/utils.go:366]: I guess `canBeCompared` may be better.

File: internal/proxy/task_search.go [hunk 13]
```diff
@@ -501,9 +501,10 @@ func (t *searchTask) PostExecute(ctx context.Context) error {
 	}
 	t.result.Results.OutputFields = t.userOutputFields
 
-	log.Debug("Search post execute done",
+	log.Info("Search post execute done",
```
Comment (longjiquan) [inline internal/proxy/task_search.go:504]: Keep `Debug` level and do not print the result since it's maybe huge.

File: internal/proxy/util.go [hunk 14]
```diff
@@ -317,6 +319,30 @@ func validateMaxLengthPerRow(collectionName string, field *schemapb.FieldSchema)
...
+func validateMaxCapacityPerRow(collectionName string, field *schemapb.FieldSchema) error {
+	exist := false
+	for _, param := range field.TypeParams {
+		if param.Key != common.MaxCapacityKey {
+			continue
+		}
+
+		maxCapacityPerRow, err := strconv.ParseInt(param.Value, 10, 64)
+		if err != nil {
+			return err
```
Comment (longjiquan) [inline internal/proxy/util.go:331]: Please clear the error message, otherwise the users may be confused by the message returned by the `strconv.ParseInt`.

**Review by longjiquan — CHANGES_REQUESTED** [review]

**Comment by longjiquan** [comment]
P.S.: Please also test the integer overflow behavior.

File: internal/core/src/common/Array.h [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+                break;
+            }
+            case ScalarArray::kIntData: {
+                element_type_ = DataType::INT32;
+                length_ = field_data.int_data().data().size();
+                data_ = new char[length_ * sizeof(int)];
+                for (int i = 0; i < length_; ++i) {
+                    reinterpret_cast<int*>(data_)[i] =
+                        field_data.int_data().data(i);
+                    offsets_.emplace_back(size_);
```
Comment (zhagnlu) [inline internal/core/src/common/Array.h:55]: vector may cost copy and relayout memory when dynamic increase, for int/float/double type:
can rewrite:
....
size_ = sizeof(T) * length_;
offsets.reserve(length_);
for(size_t i = 0; i < length_; ++i ) {
...
offsets.push_back(sizeof(T) *i);
}

File: internal/core/src/common/Array.h [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+        return element_type_;
+    }
+
+    const void*
+    data() const {
+        return data_;
+    }
+
+    bool
+    equalWithArrayValue(const proto::plan::Array& arr2) const {
```
Comment (zhagnlu) [inline internal/core/src/common/Array.h:298]: function name not follow same rule, such as print_data, equalWithArrayValue.

File: internal/core/src/common/Array.h [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+        if (arr2.array_size() != length()) {
```
Comment (zhagnlu) [inline internal/core/src/common/Array.h:299]: inner function just use length_ not need length()

File: internal/core/src/common/Array.h [hunk 18]
```diff
@@ -0,0 +1,366 @@
...
+            return false;
+        }
+        if (!arr2.same_type()) {
+            return false;
+        }
+        for (int i = 0; i < length(); i++) {
+            switch (element_type_) {
```
Comment (zhagnlu) [inline internal/core/src/common/Array.h:306]: switch out of for is better because of cpu utility

File: internal/core/src/common/VectorTrait.h [hunk 19]
```diff
@@ -52,7 +53,8 @@ constexpr bool IsVector = std::is_base_of_v<VectorTrait, T>;
 template <typename T>
 constexpr bool IsScalar =
     std::is_fundamental_v<T> || std::is_same_v<T, std::string> ||
-    std::is_same_v<T, Json> || std::is_same_v<T, std::string_view>;
+    std::is_same_v<T, Json> || std::is_same_v<T, std::string_view> ||
+    std::is_same_v<T, Array> || std::is_same_v<T, proto::plan::Array>;
```
Comment (zhagnlu) [inline internal/core/src/common/VectorTrait.h:57]: why add a proto type definition?

File: internal/core/src/mmap/Column.h [hunk 20]
```diff
@@ -286,10 +287,108 @@ class VariableColumn : public ColumnBase {
...
     std::vector<ViewType> views_{};
 };
+
+class ArrayColumn : public ColumnBase {
+ public:
+    // memory mode ctor
+    ArrayColumn(size_t num_rows,
+                const FieldMeta& field_meta,
+                const DataType element_type)
+        : ColumnBase(num_rows, field_meta), element_type_(element_type) {
```
Comment (zhagnlu) [inline internal/core/src/mmap/Column.h:302]: Could add element_type_ to filed_meta for uniform

File: internal/core/src/mmap/Column.h [hunk 21]
```diff
@@ -286,10 +287,108 @@ class VariableColumn : public ColumnBase {
...
+        for (size_t i = 0; i < indices_.size() - 1; i++) {
+            views_.emplace_back(data_ + indices_[i],
+                                indices_[i + 1] - indices_[i],
+                                element_type_,
+                                element_indices_[i]);
+        }
+        views_.emplace_back(data_ + indices_.back(),
+                            size_ - indices_.back(),
+                            element_type_,
+                            element_indices_[indices_.size() - 1]);
```
Comment (zhagnlu) [inline internal/core/src/mmap/Column.h:383]: for (size_t i = 0; i < indices_.size(); i++) {
            views_.emplace_back(data_ + indices_[i],
                               i == indices_.size() -1 ? size_ - indices_[i] : indices_[i + 1] - indices_[i],
                                element_type_,
                                element_indices_[i]);
        }
can simply code, but need  check performance

File: internal/core/src/mmap/Utils.h [hunk 22]
```diff
@@ -110,6 +111,18 @@ WriteFieldData(File& file,
                 }
                 break;
             }
+            case DataType::ARRAY: {
+                for (ssize_t i = 0; i < data->get_num_rows(); ++i) {
```
Comment (zhagnlu) [inline internal/core/src/mmap/Utils.h:115]: for num_rows, array length, these are not negative, using size_t is more suiotable,  same rule below

File: internal/core/src/query/visitors/ExecExprVisitor.cpp (hunk 8 above)
Comment (zhagnlu) [inline internal/core/src/query/visitors/ExecExprVisitor.cpp:506]: CompareTwoJsonArray,  format it

**Comment by xiaocai2333** [comment]
return ut

**Comment by xiaocai2333** [comment]
rerun ut

File: internal/core/src/common/VectorTrait.h (hunk 19 above)
Comment (xiaocai2333) [reply internal/core/src/common/VectorTrait.h:57]: because we supported the `array/json_array == [1,2,3]`, the element type is proto::plan::Array.

File: internal/core/src/common/Array.h [hunk 23]
```diff
@@ -0,0 +1,457 @@
...
+                data_ = new char[size_];
+                for (int i = 0; i < length_; ++i) {
+                    std::copy_n(field_data.string_data().data(i).data(),
+                                field_data.string_data().data(i).size(),
+                                data_ + offsets_[i]);
+                }
+                break;
+            }
+            default: {
+                //                PanicInfo("unsupported element type for array");
```
Comment (zhagnlu) [inline internal/core/src/common/Array.h:119]: throw a exception is better

File: internal/core/src/common/Array.h [hunk 24]
```diff
@@ -0,0 +1,457 @@
...
+            }
+        }
+        return true;
+    }
+
+    template <typename T>
+    T
+    get_data(const int index) const {
+        AssertInfo(
+            index >= 0 && index < length_,
```
Comment (zhagnlu) [inline internal/core/src/common/Array.h:196]: move this check to outside is better, because check to much is bad for performance

File: internal/core/src/common/Array.h [hunk 25]
```diff
@@ -0,0 +1,457 @@
...
+    bool
+    operator==(const Array& arr) const {
+        if (element_type_ != arr.element_type_) {
+            return false;
+        }
+        if (length_ != arr.length_) {
+            return false;
+        }
+        for (int i = 0; i < length_; ++i) {
+            switch (element_type_) {
```
Comment (zhagnlu) [inline internal/core/src/common/Array.h:144]: put  switch out of for is better for performance, just like
switch (element_type_) {
   case DataType::INT64:
     for (int i = 0; i< length_; ++i) {
        get_data<int64_t>(i) != arr.get_data<int64_t>(i);
     }
  .....
};

better, can use a template function to replace travel:
template<typename T>
bool TravelDataEqual(const Array& arr) {
   for (int i = 0; i< length_; ++i) {
          if( get_data<T>(i) != arr.get_data<T>(i)) {
               return false;
          }
   }
   return true;
}

File: internal/core/src/common/Array.h [hunk 26]
```diff
@@ -0,0 +1,457 @@
...
+            fmt::format(
+                "index out of range, index={}, length={}", index, length_));
+        size_t element_length = (index == length_ - 1)
+                                    ? size_ - offsets_.back()
+                                    : offsets_[index + 1] - offsets_[index];
+        if constexpr (std::is_same_v<T, std::string> ||
+                      std::is_same_v<T, std::string_view>) {
+            return T(data_ + offsets_[index], element_length);
+        }
+        if constexpr (std::is_same_v<T, int64_t> || std::is_same_v<T, double>) {
```
Comment (zhagnlu) [inline internal/core/src/common/Array.h:206]: lack of int and float type

File: internal/core/src/common/Array.h [hunk 27]
```diff
@@ -0,0 +1,457 @@
...
+                }
+                return true;
+            }
+            default:
+                return false;
+        }
+    }
+
+ private:
+    char* data_{nullptr};
```
Comment (zhagnlu) [inline internal/core/src/common/Array.h:451]: this data_ must delete it in decontruct function manually, unless will cause memory leak

File: internal/core/src/common/FieldMeta.h [hunk 28]
```diff
@@ -267,6 +294,7 @@ class FieldMeta {
     FieldName name_;
     FieldId id_;
     DataType type_ = DataType::NONE;
+    DataType element_type_ = DataType::NONE;
```
Comment (zhagnlu) [inline internal/core/src/common/FieldMeta.h:297]: name: sub_element_type is better? my personal suggest

File: internal/core/src/query/visitors/ExecExprVisitor.cpp [hunk 29]
```diff
@@ -501,6 +501,63 @@ ExecExprVisitor::ExecUnaryRangeVisitorDispatcher(UnaryRangeExpr& expr_raw)
...
+                                 std::vector<simdjson::simdjson_result<
+                                     simdjson::ondemand::value>>>) {
+        json_array_length = arr1.size();
+    }
+    if (arr2.array_size() != json_array_length) {
+        return false;
+    }
+    int i = 0;
+    for (auto&& it : arr1) {
+        switch (arr2.array(i).val_case()) {
```
Comment (zhagnlu) [inline internal/core/src/query/visitors/ExecExprVisitor.cpp:523]: same problem as up, switch out of for is better for performance

File: internal/core/src/common/Array.h (hunk 24 above)
Comment (xiaocai2333) [reply internal/core/src/common/Array.h:225]: There is no way to move it outside, because the number of elements in the array in each row may be different.

File: internal/core/src/query/visitors/ExecExprVisitor.cpp (hunk 29 above)
Comment (xiaocai2333) [reply internal/core/src/query/visitors/ExecExprVisitor.cpp:523]: This is different from the above. The elements here may be different. For example, data is: `{"x": [1, 2.2, false, "abc"]}`, the expression is `x == [1, 2.2, false, " abc"]`.

File: internal/core/src/common/FieldMeta.h (hunk 28 above)
Comment (xiaocai2333) [reply internal/core/src/common/FieldMeta.h:304]: The field element_type has already been added to proto, so for consistency, continue to call it this name.

File: internal/core/src/common/Array.h (hunk 27 above)
Comment (xiaocai2333) [reply internal/core/src/common/Array.h:423]: ok，added it.

File: internal/core/src/common/Array.h (hunk 26 above)
Comment (xiaocai2333) [reply internal/core/src/common/Array.h]: ok, added

File: internal/core/src/common/Array.h (hunk 25 above)
Comment (xiaocai2333) [reply internal/core/src/common/Array.h]: got it

File: internal/core/src/common/Array.h (hunk 23 above)
Comment (xiaocai2333) [reply internal/core/src/common/Array.h]: ok

File: internal/core/src/common/Array.h [hunk 30]
```diff
@@ -0,0 +1,429 @@
...
+        return data_array;
+    }
+
+    int
+    length() const {
+        return length_;
+    }
+
+    int
+    size() const {
```
Comment (yah01) [inline internal/core/src/common/Array.h:337]: a style suggestion: the return type of `size()` should be `size_t`, the other problem is, `size()` is usually used as the length of container (like `std::vector`, `std::map` do), so this may confuse someone, `ByteSize()` may be a good choice

**Review by yah01 — CHANGES_REQUESTED** [review]

File: internal/core/src/mmap/Column.h [hunk 31]
```diff
@@ -292,4 +293,90 @@ class VariableColumn : public ColumnBase {
...
+                            size_ - indices_.back(),
+                            element_type_,
+                            std::move(element_indices_[indices_.size() - 1]));
+    }
+
+ private:
+    std::vector<uint64_t> indices_{};
+    std::vector<std::vector<uint64_t>> element_indices_{};
+    // Compatible with current Span type
+    std::vector<Array> views_{};
```
[the remark this answers is not in the document]
Comment (yah01) [reply internal/core/src/mmap/Column.h:379]: The `ArrayView` only holds the pointer to data (which may be a mapping from `mmap()`), and never copy or free the mapping like `string_view`

File: internal/core/src/mmap/Column.h [hunk 32]
```diff
@@ -292,4 +293,98 @@ class VariableColumn : public ColumnBase {
...
+        return views_[i].output_data();
+    }
+
+    void
+    Append(const Array& array) {
+        indices_.emplace_back(size_);
+        element_indices_.emplace_back(array.get_offsets());
+        ColumnBase::Append(static_cast<const char*>(array.data()),
+                           array.byte_size());
+        views_.emplace_back(data_ + indices_.back(),
```
Comment (yah01) [inline internal/core/src/mmap/Column.h:347]: don't construct `view` here, as the later `Seal` will call `ConstructView`

**Review by yah01 — CHANGES_REQUESTED** [review]

**Comment by xiaocai2333** [comment]
rerun ut

File: internal/core/src/common/Array.h [hunk 33]
```diff
@@ -0,0 +1,639 @@
...
+                      std::is_same_v<T, std::string_view>) {
+            return T(data_ + offsets_[index], element_length);
+        }
+        if constexpr (std::is_same_v<T, int> || std::is_same_v<T, int64_t> ||
+                      std::is_same_v<T, float> || std::is_same_v<T, double>) {
+            switch (element_type_) {
+                case DataType::INT8:
+                case DataType::INT16:
+                case DataType::INT32:
+                    return static_cast<T>(reinterpret_cast<int*>(data_)[index]);
```
Comment (yah01) [inline internal/core/src/common/Array.h:253]: reinterpret to `int32_t`, the size of `int` may be 64 bits, not always 32 bits on all platforms.

The other codes like this should be also modified

File: internal/core/src/mmap/Column.h [hunk 34]
```diff
@@ -327,4 +328,89 @@ class VariableColumn : public ColumnBase {
...
+        for (size_t i = 0; i < indices_.size() - 1; i++) {
+            views_.emplace_back(data_ + indices_[i],
+                                indices_[i + 1] - indices_[i],
+                                element_type_,
+                                std::move(element_indices_[i]));
+        }
+        views_.emplace_back(data_ + indices_.back(),
+                            size_ - indices_.back(),
+                            element_type_,
+                            std::move(element_indices_[indices_.size() - 1]));
```
Comment (yah01) [inline internal/core/src/mmap/Column.h:406]: let's clear the `element_indices_` as it's no longer needed at the end

**Review by yah01 — CHANGES_REQUESTED** [review]
